### PR TITLE
[blog] benchmarks: a cautionary tale

### PR DESCRIPTION
**Pull request by bennn**
Short story on some trouble we had with unreliable numbers due to shared memory contention. We thought we had an embarrasingly-parallelizable job, and got a little eager collecting data before we checked it.

Thanks to Jan, for prompting me at ECOOP to organize this story.

To anyone free to review:
- is there enough detail -- do you understand the graphs?
- any open questions at the end?
  I'd like to keep this short, so let me know about boring parts too.

### CODE REVIEW

**Comment by gasche** [comment]
Is there a HTML rendering of the draft somewhere? I think it would help as the post relies on the graphs in context.

**Comment by bennn** [comment]
Here's a simulated PRL blog experience:
http://www.ccs.neu.edu/home/types/proto-blog/

**Comment by gasche** [comment]
I like the post, it is cheerful and nice to read. I still have a long list
of minor nitpicks!

> this is the only sensible way to measure the performance of a gradual
> type system

I think you could be more careful and say "this is the only realistic way
to measure the performance of a gradual type system that we know of."

> As good scientists, we explored practical and theoretical solutions.

I'm not sure everyone would see the humor in the "As good scientists" part,
you could just drop it.

> put it to work running benchmarks on 29 cores simultaneously

There are 32 cores on the machine. Assuming you leave one core for the OS
and master thread, that leaves two cores unaccounted for. Out of curiosity,
why 29?

> Fast-forward to March, it was time to test the prediction algorithm.
> Again using 29 cores (because, why not), we gathered baseline numbers for
> one benchmark then measured the costs and benefits of interactions between
> modules in the benchmark. For instance, [...]

I started reading this paragraph as if I was going to hear a performance
story, and by the time I noticed that you were actually giving a
description of the non-trivial prediction algorithm I realized that I was a
bit lost and confused. I think it would be better to give a full paragraph
to explaining the science behind prediction work (maybe a bit more
detailed, so that one can follow), and only then tell about the
measurements for it. (For example, one needs to understand that the
prediction relies on running a subset of the configuration to understand
that they also rely on measurements.)

> Except for configuration 9

The close match I see on the figure is configuration 8 rather than 9.

> The hope was that results taken from 1 core could match results from T
> cores

Could you use as meta-variable a letter that looks less like "1"? I think
"N" would be fine.

> And so, we said goodbye to the last 4 months of data and started over
> running at most two cores. The new results are all stable, but still we
> keep pinching ourselves.

It might be useful to mention that the results of the POPL paper are not in
any way contradicted by this new data.

**Comment by bennn** [comment]
Thanks very much! 

It was 29 "just to be safe". We really, really should have just tested how the `#` cores affected performance at the beginning.

**Comment by bennn** [comment]
Updated the PR and [link](http://www.ccs.neu.edu/home/types/proto-blog/), tho from my end the 3rd image on the link isn't rendering. Will check again at home.

@maxsnew can you OK the paragraph on the prediction algorithm?
https://github.com/nuprl/nuprl.github.io/pull/49/files#diff-c357616cf903cf765779330d8fffcf5cR42

File: blog/_src/posts/2016-08-03-a-few-cores-too-many.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+The next question we asked was naturally "how can we scale this method to large software projects?"
+In [our case](http://docs.racket-lang.org/ts-reference/Libraries_Provided_With_Typed_Racket.html#%28part._.Porting_.Untyped_.Modules_to_.Typed_.Racket%29),
+ the number of gradually typed configurations scaled exponentially with the number of modules.
+Current gradual type system for [Python](https://github.com/mvitousek/reticulated)
+ and [JavaScript](http://www.di.ens.fr/~zappa/readings/ecoop15.pdf)
+ are exponential in the number of _variables_ in the program.
+
+We explored two solutions to the problem, one theoretical, one practical.
+
+  1. Max New began work on an additive predication model (inspired by work
```
Comment (maxsnew) [inline blog/_src/posts/2016-08-03-a-few-cores-too-many.md]: prediction model

File: blog/_src/posts/2016-08-03-a-few-cores-too-many.md [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+With benchmarks running, Max implemented a prediction algorithm.
+The basic idea was to focus on _boundaries_ between modules and isolate their
+ effect on performance.
+If two modules are untyped, their boundary will have zero cost.
+If the same two modules are typed, their boundary might result in an overall
+ performance improvement (decrease) due to type-driven optimizations.
+And if one module is typed and the other untyped, their boundary will
+ suffer some cost of type checking at runtime.
+
+Fast-forward to March, we had a prototype predication algorithm and it was time to test.
```
Comment (maxsnew) [inline blog/_src/posts/2016-08-03-a-few-cores-too-many.md]: prediction

**Comment by maxsnew** [comment]
It's fine besides the typo "predication" a couple times.

Maybe just add one sentence there making it a bit more obvious why this would reduce from exponential to linear/quadratic

**Comment by bennn** [comment]
Typos fixed, and added a sentence on exponential vs. quadratic.
Planning to post this evening.

**Comment by MiloDavis** [comment]
It might help if you explain which benchmark the graphs are showing, I'm not sure if it's one benchmark or multiple.  Other than that, it looks good to me.

**Comment by bennn** [comment]
thanks all!

**Pull request merged**